spec/ordinary-and-exotic-object-behaviours.html

```diff
@@ -0,0 +1,90 @@
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. Throw a *TypeError* exception.
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-record-exotic-objects-ownpropertykeys">
+        <h1>[[OwnPropertyKeys]] ( )</h1>
+        <p>When the [[OwnPropertyKeys]] internal method of a Record exotic object _R_ is called, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _keys_ be a new empty List.
+          1. Let _rec_ be _R_.[[RecordData]].
+          1. Assert: Type(_rec_) is Record.
+          1. Let _len_ be the length of _rec_.
+          1. For each primitive data property _pdp_, do
+            1. Let _key_ be the String key value of _pdp_
+            1. Assert: Type(_key_) is String
+            1. Add _key_ to _keys_
+          1. Return _keys_
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-recordcreate" aoid="RecordCreate">
+        <h1>RecordCreate ( _value_ )</h1>
+        <p>This abstract operation RecordCreate takes a _value_ argument. It is used to specify the creation of new Record exotic objects. It performs the following steps when called:</p>
+        <emu-alg>
+            1. Assert: Type(_value_) is Record.
+            1. Let _R_ be ! MakeBasicObject(&laquo; [[RecordData]] &raquo;).
+            1. Set _R_.[[RecordData]] to _value_
+            1. Set _R_.[[GetOwnProperty]] as specified in <emu-xref href="#sec-record-exotic-objects-getownproperty-p"></emu-xref>.
+            1. Set _R_.[[DefineOwnProperty]] as specified in <emu-xref href="#sec-record-exotic-objects-defineownproperty-p-desc"></emu-xref>.
+            1. Set _R_.[[OwnPropertyKeys]] as specified in <emu-xref href="#sec-record-exotic-objects-ownpropertykeys"></emu-xref>.
+            1. Return _R_
+            
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-recordgetownproperty-r-p" aoid="RecordGetOwnProperty">
+        <h1>RecordGetOwnProperty ( _R_, _P_ )</h1>
+        <p>This abstract operation RecordGetOwnProperty takes arguments _R_ and _P_. It performs the following steps when called:</p>
+        <emu-alg>
+            1. Assert: _R_ is an Object that has [[RecordData]] internal slot.
+            1. Assert: Type(_P_) is String.
+            1. Let _rec_ be _R_.[[RecordData]]
+            1. Assert Type(rec) is Record
+            1. Let _len_ be the length of _rec_.
+            1. Let _resultPrim_ be *undefined*
+            1. For each primitive data property _pdp_, do
+              1. Let _key_ be the String key value of _pdp_
+              1. Let _prim_ be the associated value of _pdp_
+              1. Assert: Type(_key_) is String
+              1. Assert: Type(_prim_) is not Object
+              1. If _key_ == _P_, let _resultPrim_ be _prim_
+            1. If _resultPrim_ is *undefined*, return *undefined*
```

Comment:
It should be possible to have a data property of a Record whose value is *undefined*, so this seems a little broken. Maybe you could start _resultPrim_ as ~empty~ and check for that here instead.